python_modules/dagster/dagster/core/asset_defs/asset_collection.py

```diff
@@ -68,6 +71,115 @@ def all_assets_job_name(self) -> str:
+                else:
+                    excluded_assets.append(asset)
+        else:
+            included_assets = cast(List[AssetsDefinition], self.assets)
+            excluded_assets = list(self.source_assets)
+
+        return build_assets_job(
+            name=name,
+            assets=included_assets,
+            source_assets=excluded_assets,
+            resource_defs=self.resource_defs,
+            executor_def=self.executor_def,
+            description=description,
+            tags=tags,
+        )
+
+    def _parse_asset_selection(self, selection, job_name) -> List[str]:
+        """Convert selection over asset key to selection over ops"""
+
+        asset_keys_to_ops: Dict[str, List[OpDefinition]] = {}
+
+        if isinstance(selection, str):
+            selection = [selection]
+
+        if len(selection) == 1 and selection[0] == "*":
+            return selection
+
+        source_asset_keys = set()
+        for asset in self.assets:
+            for asset_key in asset.asset_keys:
+                asset_key_as_str = ".".join([piece for piece in asset_key.path])
+                if not asset_key_as_str in asset_keys_to_ops:
+                    asset_keys_to_ops[asset_key_as_str] = []
+                asset_keys_to_ops[asset_key_as_str].append(asset.op)
+
+        for asset in self.source_assets:
+            if isinstance(asset, ForeignAsset):
+                asset_key_as_str = ".".join([piece for piece in asset.key.path])
+                source_asset_keys.add(asset_key_as_str)
+            else:
+                for asset_key in asset.asset_keys:
+                    asset_key_as_str = ".".join([piece for piece in asset_key.path])
+                    source_asset_keys.add(asset_key_as_str)
+
+        op_selection = []
+
+        for clause in selection:
+            token_matching = re.compile(r"^(\*?\+*)?([.\w\d\[\]?_-]+)(\+*\*?)?$").search(
+                clause.strip()
+            )
+            # return None if query is invalid
+            parts = token_matching.groups() if token_matching is not None else None
+            if parts is None:
+                raise DagsterInvalidDefinitionError(
+                    f"When attempting to create job '{job_name}', the clause {clause} within the asset key selection was invalid. Please review the selection syntax here (imagine there is a link here to the docs)."
+                )
+            upstream_part, key_str, downstream_part = parts
+            if key_str in source_asset_keys:
+                raise DagsterInvalidDefinitionError(
+                    f"When attempting to create job '{job_name}', the clause '{clause}' selects asset_key '{key_str}', which comes from a source asset. Source assets can't be materialized, and therefore can't be subsetted into a job. Please choose a subset on asset keys that are materializable - that is, included on assets within the collection. Valid assets: {list(asset_keys_to_ops.keys())}"
```

Comment:
These are really long lines. Can we spread these error messages out over multiple lines?

Reply:
ah my bad my word wrap makes me oblivious to this pain